---
title: "Differential abundance analysis of Seasol metagenomic data"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    fig_width: 7
    fig_height: 7
theme: cosmo
---

Source: TBA

## Intro

Here we are determining whether there is any differentially abundant microbes in these
soil samples with and without treatment of Seasol.

```{r,begin}

library("DESeq2")
library(mitch)

```


## Load and clean data

```{r,load1}

x <- readLines("absoluteabundance.txt")
xx <- strsplit(x,"\t")
df <- as.data.frame(do.call(rbind, xx))
colnames(df) <- df[1,]
df <- df[2:nrow(df),]

x <- df[,grep("kraken",colnames(df))]
x <- apply(x,2,as.numeric)
x <- as.data.frame(x)
rownames(x) <- df$`#OTU ID`
colnames(x) <- gsub("_kraken_bracken","",colnames(x))

nrow(x)
x <- x[which(rowMeans(x)>10),]
nrow(x)

```


## MDS

```{r,mds1}

cols <- as.numeric(grepl("S",colnames(x)))
cols <- gsub("0","lightblue",cols) 
cols <- gsub("1","pink",cols) 

plot(cmdscale(dist(t(x))),
  xlab="Coordinate 1", ylab="Coordinate 2", col=cols,
  pch=19,cex=4, main="MDS plot")
text(cmdscale(dist(t(x))), labels=colnames(x) ) 

```


## DESeq2

```{r,deseq1}

ss <- as.data.frame(colnames(x))
ss$s <- factor(as.numeric(grepl("S",colnames(x))))

y <- DESeqDataSetFromMatrix(countData = round(x), colData = ss, design = ~ s)
y <- DESeq(y)
res <- results(y)
res <-as.data.frame(res[order(res$pvalue),])

head(res,10)

```

## Charts

```{r,deplots}

maplot <- function(de,contrast_name) {
  sig <-subset(de, padj < 0.05 )
  up <-rownames(subset(de, padj < 0.05 & log2FoldChange > 0))
  dn <-rownames(subset(de, padj < 0.05 & log2FoldChange < 0))
  GENESUP <- length(up)
  GENESDN <- length(dn)
  DET=nrow(de)
  SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down", DET, "detected")
  ns <-subset(de, padj > 0.05 )
  plot(log2(de$baseMean),de$log2FoldChange, 
       xlab="log2 basemean", ylab="log2 foldchange",
       pch=19, cex=0.5, col="dark gray",
       main=contrast_name, cex.main=0.7)
  points(log2(sig$baseMean),sig$log2FoldChange,
         pch=19, cex=0.5, col="red")
  mtext(SUBHEADER,cex = 0.7)
}
make_volcano <- function(de,name) {
    sig <- subset(de,padj<0.05)
    N_SIG=nrow(sig)
    N_UP=nrow(subset(sig,log2FoldChange>0))
    N_DN=nrow(subset(sig,log2FoldChange<0))
    DET=nrow(de)
    HEADER=paste(N_SIG,"@5%FDR,", N_UP, "up", N_DN, "dn", DET, "detected")
    plot(de$log2FoldChange,-log10(de$padj),cex=0.5,pch=19,col="darkgray",
        main=name, xlab="log2 FC", ylab="-log10 pval")
    mtext(HEADER)
    grid()
    points(sig$log2FoldChange,-log10(sig$padj),cex=0.5,pch=19,col="red")
}
maplot(res,"ctrl vs seasol")
make_volcano(res,"ctrl vs seasol")

```

## Identify OTUs

```{r,OTU ident}

otu <- df[,c(1,ncol(df))]
rownames(otu) <- otu[,1]
otu[,1] = NULL

resm <- merge(res,otu,by=0)
resm <-as.data.frame(resm[order(resm$pvalue),])

head(resm,10)

# library size normalised
x <- x+1
xx <- x/colSums(x)*1000000

xx <- xx[,order(colnames(xx)),]

top_otus <- head(resm,10)$Row.names

par(mfrow=c(5,2))
bars <- sapply(top_otus,function(o) {
  vals <- as.numeric(xx[which(rownames(xx) == o),])
  names(vals) <- colnames(xx)
  taxon <- otu[which(rownames(otu) == o),]
  barplot(vals,ylab="log2 CPM")
  mtext(taxon,cex=0.4)
  grid()
})

```

## Taxonomic dataframe

Figure out the main groups of microbes for enrichment analysis

```{r,taxdf1}

tax <- strsplit(otu$`Consensus Lineage`,";")
taxdf <- as.data.frame(do.call(rbind, tax))
colnames(taxdf) <- c("kingdom","phylum","class","order","family","genus","species")
rownames(taxdf) <- rownames(otu)
taxdf <- taxdf[which(rownames(taxdf) %in% rownames(res)),]
apply(taxdf,2,function(y){ length(unique(y))})


```

## Enrichment analysis using tanonomic levels

### Making OTU sets

```{r,enrich1}

phylums <- unique(taxdf$phylum)
phylum_sets <- lapply(phylums, function(group){
  rownames(taxdf[which(taxdf$phylum==group),])
})
names(phylum_sets) <- phylums
phylum_sets <- phylum_sets[lapply(phylum_sets,length)>=5]


classes <- unique(taxdf$class)
class_sets <- lapply(classes, function(group){
  rownames(taxdf[which(taxdf$class==group),])
})
names(class_sets) <- classes
class_sets <- class_sets[lapply(class_sets,length)>=5]

orders <- unique(taxdf$order)
order_sets <- lapply(orders, function(group){
  rownames(taxdf[which(taxdf$order==group),])
})
names(order_sets) <- orders
order_sets <- order_sets[lapply(order_sets,length)>=5]

familys <- unique(taxdf$family)
family_sets <- lapply(familys, function(group){
  rownames(taxdf[which(taxdf$family==group),])
})
names(family_sets) <- familys
family_sets <- family_sets[lapply(family_sets,length)>=5]

genuss <- unique(taxdf$genus)
genus_sets <- lapply(genuss, function(group){
  rownames(taxdf[which(taxdf$genus==group),])
})
names(genus_sets) <- genuss
genus_sets <- genus_sets[lapply(genus_sets,length)>=5]

```

### Enrichment with mitch

```{r,mitch1}

m <- mitch_import(res,DEtype="DESeq2")
mphylum <- mitch_calc(m,genesets=phylum_sets)
head(mphylum$enrichment_result,10)
mitch_report(res=mphylum,outfile="mitch_phylum.html",overwrite=TRUE)

mclass <- mitch_calc(m,genesets=class_sets)
head(mclass$enrichment_result,10)
mitch_report(res=mclass,outfile="mitch_class.html",overwrite=TRUE)

morder <- mitch_calc(m,genesets=order_sets)
head(morder$enrichment_result,20)
mitch_report(res=morder,outfile="mitch_order.html",overwrite=TRUE)

mfamily <- mitch_calc(m,genesets=family_sets)
head(mfamily$enrichment_result,20)
mitch_report(res=mfamily,outfile="mitch_family.html",overwrite=TRUE)

mgenus <- mitch_calc(m,genesets=genus_sets)
head(mgenus$enrichment_result,20)
mitch_report(res=mgenus,outfile="mitch_genus.html",overwrite=TRUE)

```

## Session information

```{r,sessioninfo}

sessionInfo()

```
